pkg/antctl/raw/upgrade/apistorage/command.go

```diff
@@ -0,0 +1,288 @@
+func getCRDsToUpgrade(writer io.Writer, k8sClient client.Client, crdNameListToUpgrade sets.Set[string]) []*apiextv1.CustomResourceDefinition {
+	var crdsToUpgrade []*apiextv1.CustomResourceDefinition
+	for crdName := range crdNameListToUpgrade {
+		crd := &apiextv1.CustomResourceDefinition{}
+		if err := k8sClient.Get(context.TODO(), client.ObjectKey{Name: crdName}, crd); err != nil {
+			fmt.Fprintf(writer, "Failed to get CRD %q: %v.\n", crdName, err)
+			continue
+		}
+		// Skip the CRD that has only one version.
+		if len(crd.Spec.Versions) == 1 {
+			fmt.Fprintf(writer, "Skip upgrading CRD %q since it only has one version.\n", crdName)
+			continue
+		}
+		// Skip the CRD that all stored objects are in the latest version.
+		if len(crd.Status.StoredVersions) == 1 && crd.Status.StoredVersions[0] == getCRDStorageVersion(crd) {
+			fmt.Fprintf(writer, "Skip upgrading CRD %q since all stored objects are in the latest version.\n", crdName)
+			continue
+		}
+		crdsToUpgrade = append(crdsToUpgrade, crd)
+	}
+	return crdsToUpgrade
+}
+
+// upgradeCRDObjects upgrades the existing objects of a CRD.
+func upgradeCRDObjects(writer io.Writer, k8sClient client.Client, crd *apiextv1.CustomResourceDefinition) error {
+	objList := &unstructured.UnstructuredList{}
+	objList.SetGroupVersionKind(schema.GroupVersionKind{
+		Group:   crd.Spec.Group,
+		Version: getCRDStorageVersion(crd),
+		Kind:    crd.Spec.Names.ListKind,
+	})
+
+	if err := k8sClient.List(context.TODO(), objList); err != nil {
+		return err
+	}
+
+	if len(objList.Items) == 0 {
+		return nil
+	}
+
+	fmt.Fprintf(writer, "Upgrading %d objects of CRD %q.\n", len(objList.Items), crd.Name)
+	failed := 0
+	for _, item := range objList.Items {
+		if err := upgradeCRDObject(writer, k8sClient, crd, item); err != nil {
+			fmt.Fprintf(writer, "Failed to upgrade object %s/%s of CRD %q.\n", item.GetNamespace(), item.GetName(), crd.Name)
+			failed++
+		}
+	}
+	fmt.Fprintf(writer, "Successfully upgraded %d objects of CRD %q.\n", len(objList.Items)-failed, crd.Name)
+
+	return nil
+}
+
+func upgradeCRDObject(writer io.Writer, k8sClient client.Client, crd *apiextv1.CustomResourceDefinition, item unstructured.Unstructured) error {
+	return retry.OnError(wait.Backoff{
+		Duration: time.Second,
+		Steps:    5},
+		// If the error is that the object is not found, don't retry; retry for other errors.
+		func(err error) bool {
+			if apierrors.IsNotFound(err) {
+				fmt.Fprintf(writer, "Skip upgrading object %s/%s of CRD %q which is not found.\n", item.GetNamespace(), item.GetName(), crd.Name)
+				return false
+			}
+			return true
+		},
+		// Get the object for every retry.
+		func() error {
+			key := client.ObjectKeyFromObject(&item)
+			obj := &unstructured.Unstructured{}
+			obj.SetGroupVersionKind(schema.GroupVersionKind{
+				Group:   crd.Spec.Group,
+				Version: getCRDStorageVersion(crd),
+				Kind:    crd.Spec.Names.Kind,
+			})
+			if err := k8sClient.Get(context.TODO(), key, obj); err != nil {
```

Comment:
Shouldn't it only get the object after updating fails with conflict error? otherwise it could waste time on getting each object once?

Reply:
Thanks for reminding that, done as you said.